### PR TITLE
refactor(core): Replace `pathtree.Tree()` by alternatives

### PR DESCRIPTION
**Pull request by timoffex**
Description
---
Removes `pathtree.Tree()` which is dangerous to use and cleans up all former usages of it.

A few places are modified to use `CloneTree()`, some others `GetLeaf()`, and I also added `ToExtendedJSON()` for serialization use-cases.

### CODE REVIEW

File: core/internal/pathtree/pathtree.go [hunk 1]
```diff
@@ -1,5 +1,7 @@
 package pathtree
 
+import "github.com/wandb/segmentio-encoding/json"
```
Comment (kptkin) [inline core/internal/pathtree/pathtree.go:3]: i assume you will remove it in your follow up PR right?

File: core/internal/runhistory/runhistory.go [hunk 2]
```diff
@@ -121,9 +115,27 @@ func (rh *RunHistory) Flatten() ([]*service.HistoryItem, error) {
 	return history, nil
 }
 
-// Clone returns a deep copy of the history tree.
-func (rh *RunHistory) Tree() pathtree.TreeData {
-	return rh.pathTree.Tree()
+// GetNumber returns the value of a number-valued metric.
+func (rh *RunHistory) GetNumber(path ...string) (float64, bool) {
```
Comment (kptkin) [inline core/internal/runhistory/runhistory.go:119]: i don't know if want to mess with user provided types, if we are doing it already in the code it's not ideal

File: core/internal/runhistory/runhistory.go [hunk 3]
```diff
@@ -121,9 +115,27 @@ func (rh *RunHistory) Flatten() ([]*service.HistoryItem, error) {
...
+	value, exists := rh.pathTree.GetLeaf(pathtree.TreePath(path))
+	if !exists {
+		return 0, false
+	}
+
+	switch x := value.(type) {
+	case int64:
```
Comment (kptkin) [inline core/internal/runhistory/runhistory.go:126]: do you need to handle any other ints?

File: core/internal/runhistory/runhistory.go [hunk 4]
```diff
@@ -121,9 +115,27 @@ func (rh *RunHistory) Flatten() ([]*service.HistoryItem, error) {
...
+		return float64(x), true
+	case float64:
```
Comment (kptkin) [inline core/internal/runhistory/runhistory.go:128]: same as above any other floats?

File: core/internal/runhistory/runhistory.go (hunk 2 above)
Comment (kptkin) [reply core/internal/runhistory/runhistory.go:119]: oh is it only used for testing?

**Review by kptkin — APPROVED** [review]

File: core/internal/pathtree/pathtree.go (hunk 1 above)
Comment (timoffex) [reply core/internal/pathtree/pathtree.go:3]: yes---all these PRs are work I ended up having to do to remove it while minimizing unexpected breakages

File: core/internal/runhistory/runhistory.go (hunk 2 above)
Comment (timoffex) [reply core/internal/runhistory/runhistory.go:119]: yeah, seems like only for testing

JSON numbers are guaranteed to unmarshal to float64 in encoding/json by the way

File: core/internal/runhistory/runhistory.go (hunk 3 above)
Comment (timoffex) [reply core/internal/runhistory/runhistory.go:126]: no---int64 and float64 are the only possibilities. With encoding/json (and segmentio) it's always float64 actually---but I'm preparing for my simplejsonext PR which can also unmarshal to int64.

File: core/internal/runhistory/runhistory.go (hunk 2 above)
Comment (timoffex) [reply core/internal/runhistory/runhistory.go]: removed it

File: core/internal/runhistory/runhistory.go (hunk 3 above)
Comment (kptkin) [reply core/internal/runhistory/runhistory.go]: got it thanks!

**Pull request merged**